### PR TITLE
[VAULT-2776] Add prefix_filter option to Vault

### PR DESCRIPTION
**Pull request by pmmukh**
- Adds a `prefix_filter` option to the telemetry configs, that can explicitly allow or block certain metric prefixes.
- Adds a `filter_default` option to the telemetry configs, which is required in conjunction with the config above, to decide whether all metrics are allowed or denied by default. This is currently always true, but users may want to control this in their vault config along with `prefix_filter`.
- Updates documentation to cover both configs

Note that this attempts to maintain parity with this config in consul https://www.consul.io/docs/agent/options#telemetry-prefix_filter
Also, the logic for parsing the prefixFilter config is pretty much lifted from here https://github.com/hashicorp/consul/blob/main/agent/config/builder.go#L640-L654

Related GH issue: https://github.com/hashicorp/vault/issues/3792

### CODE REVIEW

File: internalshared/configutil/telemetry.go [hunk 1]
```diff
@@ -388,5 +401,29 @@ func SetupTelemetry(opts *SetupTelemetryOpts) (*metrics.InmemSink, *metricsutil.
...
+			log.Print("Cannot have empty filter rule in prefix_filter")
+			continue
+		}
+		switch rule[0] {
+		case '+':
+			telemetryAllowedPrefixes = append(telemetryAllowedPrefixes, rule[1:])
+		case '-':
+			telemetryBlockedPrefixes = append(telemetryBlockedPrefixes, rule[1:])
+		default:
+			log.Printf("Filter rule must begin with either '+' or '-': %q", rule)
```
Comment (ncabatoff) [inline internalshared/configutil/telemetry.go:425]: We never use the stdlib `log` in Vault.  I would add an `error` return value to this func instead.

File: website/content/docs/configuration/telemetry.mdx [hunk 2]
```diff
@@ -51,6 +51,17 @@ The following options are available on all telemetry configurations.
 - `add_lease_metrics_namespace_labels` `(bool: false)` - If this value is set to true, then `vault.expire.leases.by_expiration`
   will break down expiring leases by both time and namespace. This parameter is disabled by default because enabling it can lead
   to a large-cardinality metric.
+- `filter_default` - This controls whether to allow metrics that have not been specified by the filter.
```
Comment (ncabatoff) [inline website/content/docs/configuration/telemetry.mdx:54]: Please add `(bool: false)` to describe this param.

File: internalshared/configutil/telemetry_test.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParsePrefixFilters(t *testing.T) {
+	prefixFilters := []string{"", "+vault.abc", "-vault.abc", "vault.abc"}
+
+	allowedPrefixes, blockedPrefixes := parsePrefixFilter(prefixFilters)
+
+	assert.Equal(t, len(allowedPrefixes), 1)
```
Comment (ncabatoff) [inline internalshared/configutil/telemetry_test.go:14]: You can say `assert.Len` to check the length with testify.

File: internalshared/configutil/telemetry.go (hunk 1 above)
Comment (pmmukh) [reply internalshared/configutil/telemetry.go:425]: Ah yeah I wasn't sure if I should use `log` here. I think the only reason I didn't do error return, is cause the equivalent consul code logs out a warning when the filter is improperly formatted versus failing https://github.com/hashicorp/consul/blob/main/agent/config/builder.go#L652, so I was looking to maintain that pattern. Maybe `opts.UI.Warn` would make sense https://github.com/hashicorp/vault/blob/main/internalshared/configutil/telemetry.go#L371 ?

**Comment by ncabatoff** [comment]
I know Consul doesn't treat a config error here as a failure and will start up with warnings, but I would prefer we were stricter.  I think it's a bad UX to ignore config errors.

Also, could you add fixture-based test(s) like we do for most other config parsing code?  Look under `command/server` for some examples.

**Comment by pmmukh** [comment]
> I know Consul doesn't treat a config error here as a failure and will start up with warnings, but I would prefer we were stricter. I think it's a bad UX to ignore config errors.
Roger, i would concur with that tbh, will change it to error out.

> Also, could you add fixture-based test(s) like we do for most other config parsing code? Look under `command/server` for some examples.
Will do!

File: internalshared/configutil/telemetry.go (hunk 1 above)
Comment (pmmukh) [reply internalshared/configutil/telemetry.go]: 6a28af2

File: website/content/docs/configuration/telemetry.mdx (hunk 2 above)
Comment (pmmukh) [reply website/content/docs/configuration/telemetry.mdx]: 6a28af2

File: internalshared/configutil/telemetry_test.go (hunk 3 above)
Comment (pmmukh) [reply internalshared/configutil/telemetry_test.go]: I ended up just doing an `assert.Equal` against an expected array, while refactoring the test
6a28af2

**Comment by pmmukh** [comment]
> Also, could you add fixture-based test(s) like we do for most other config parsing code? Look under `command/server` for some examples.

6a28af2

File: internalshared/configutil/telemetry_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		},
+	}
+	t.Run("validate metric filter configs", func(t *testing.T) {
+		t.Parallel()
+
+		for _, tc := range cases {
+
+			allowedPrefixes, blockedPrefixes, err := parsePrefixFilter(tc.inputFilters)
+
+			if err != nil {
```
Comment (ncabatoff) [inline internalshared/configutil/telemetry_test.go:43]: It's fine with the cases you have now, but I consider it best practice when you have an expected error field in your test case type to do something like

if err != nil { compare with expected error }
else { assert expected error empty}

File: command/server/config_telemetry_test.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+func TestMetricFilterConfigs(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		configFile            string
+		expectedFilterDefault *bool
+		expectedPrefixFilter  []string
+	}{
+		{
+			"./test-fixtures/telemetry/valid_prefix_filter.hcl",
```
Comment (ncabatoff) [inline command/server/config_telemetry_test.go:17]: Forgot to add fixtures?

Comment (pmmukh) [reply command/server/config_telemetry_test.go:17]: hah yep just noticed that I didn't push them, looks like the dir I created them in got gitignored, fixing that

Comment (pmmukh) [reply command/server/config_telemetry_test.go:17]: 6c2ad99

File: internalshared/configutil/telemetry_test.go (hunk 4 above)
Comment (pmmukh) [reply internalshared/configutil/telemetry_test.go:43]: ah that makes sense! added that here 6c2ad99

File: .gitignore [hunk 6]
```diff
@@ -48,7 +48,7 @@ Vagrantfile
 # Configs
 *.hcl
 !command/agent/config/test-fixtures/*.hcl
-!command/server/test-fixtures/*.hcl
+!command/server/test-fixtures/*/*.hcl
```
Comment (ncabatoff) [inline .gitignore:51]: Should this be `**/*.hcl` instead of `*/*.hcl`?

**Review by ncabatoff — APPROVED** [review]
Looks good, possibly modulo the .gitignore change.

File: .gitignore (hunk 6 above)
Comment (pmmukh) [reply .gitignore]: it should! beae728

**Pull request merged**